frodo.js

```diff
@@ -1,42 +1,39 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?